src/twisted/protocols/_smb/base.py

```diff
@@ -0,0 +1,207 @@
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+# -*- test-case-name: twisted.protocols._smb.tests -*-
+"""
+base classes for SMB networking
+"""
+
+import struct
+
+from twisted.internet import protocol
+from twisted.logger import Logger
+from twisted.python.randbytes import secureRandom
+
+log = Logger()
+
+
+
+class SMBError(Exception):
+    """SMB specific errors
+    """
+    def __init__(self, msg, ntstatus=0xC0000001):
+        self.msg = msg
+        self.ntstatus = ntstatus
+
+    def __str__(self):
+        return "%s 0x%08x" % (self.msg, self.ntstatus)
+
+
+
+def u2nt_time(epoch):
+    """
+    Convert UNIX epoch time to NT filestamp
+    quoting from spec: The FILETIME structure is a 64-bit value
+    that represents the number of 100-nanosecond intervals that
+    have elapsed since January 1, 1601, Coordinated Universal Time.
+
+    @param epoch: seconds since 1/1/1970
+    @type epoch: L{float}
+
+    @rtype: L{int}
+    """
+    return int(epoch * 10000000.0) + 116444736000000000
+
+
+
+class GeneralStruct:
+    """
+    a more ergonomic struct.unpack/pack. an object that represents an ordered
+    binary structure and field names..
+    field names are acceessible as attributes.
+    don't use directly, subclass through L{base.nstruct}
+    """
+    def __init__(self, data=None):
+        """
+        @param data: binary data sent to C{pack}
+        @type data: L{bytes}
+        """
+        self.buffer = b''
+        self._values = {}
+        if data:
+            self.unpack(data)
+
+    def __setattr__(self, name, val):
+        if name in self._fields_names:
+            self._values[name] = val
+        else:
+            object.__setattr__(self, name, val)
+
+    def zero(self, field):
+        """
+        get logical zero/null value for a field
+        """
+        n = self._fields_names.index(field)
+        t = self._types[n]
+        if t.endswith("s"):
+            return b"\0" * int(t[0:-1])
+        else:
+            return 0
+
+    def __getattr__(self, name):
+        if name in self._fields_names:
+            return self._values.get(name, self.zero(name))
+        else:
+            raise AttributeError(name)
+
+    def pack(self):
+        """
+        Pack data according to the pattern
+
+        @rtype: L{bytes}
+        """
+        p = [self._values.get(n, self.zero(n)) for n in self._fields_names]
+        return struct.pack(self._pattern, *p) + self.buffer
+
+    def unpack(self, data):
+        """
+        unpack data according to the pattern
+        exceas data goes in buffer
+
+        @param data: the data
+        @type data: L{bytes}
+        """
+        ret = struct.unpack(self._pattern, data[:self._size])
+        for i in range(len(ret)):
+            self._values[self._fields_names[i]] = ret[i]
+        self.buffer = data[self._size:]
```

Comment:
It really looks like you want to be using a `struct.Struct` here rather than the function versions

Reply:
new implementation does use Struct